src/bson_mapper/bson_streambuf.cpp

```diff
@@ -0,0 +1,86 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cassert>
+#include <cstdio>
+#include <iostream>
+#include <streambuf>
+
+#include <mongocxx/client.hpp>
+#include <mongocxx/instance.hpp>
+
+#include "bson_streambuf.hpp"
+
+#include <bson_mapper/config/prelude.hpp>
+
+namespace bson_mapper {
+BSON_MAPPER_INLINE_NAMESPACE_BEGIN
+
+bson_output_streambuf::bson_output_streambuf(mongocxx::collection *coll) {
+    this->coll = coll;
+}
+
+bson_output_streambuf::~bson_output_streambuf() {
+    if (data != nullptr) {
+        delete[] data;
+    }
+}
+
+int bson_output_streambuf::underflow() {
+    return EOF;
+}
+
+int bson_output_streambuf::overflow(int ch) {
+    int result = EOF;
+    assert(ch >= 0 && ch <= UCHAR_MAX);
+    result = insert(ch);
+    return result;
+}
+
+int bson_output_streambuf::insert(int ch) {
+    bytes_read++;
+
+    // for first four bytes, build int32 that contains document size
+    if (bytes_read <= 4) {
+        len += (ch << (8 * (bytes_read - 1)));
+    }
+    // once doc size is received, allocate space.
+    if (bytes_read == 4) {
+        // TODO perhaps fail unless 0 < len <= MAX_BSON_SIZE?
+        data = new uint8_t[len];
+        *(size_t *)data = len;
+    }
+
+    if (bytes_read > 4) {
+        data[bytes_read - 1] = (uint8_t)ch;
```

Comment:
We try to avoid c-style casts. Can this be done with a static_cast?
e.g. `static_cast<uint8_t>(ch)`?